### PR TITLE
Tuple type always parses a minimum of two elements

### PR DESCRIPTION
**Pull request by jcouv**
**Customer scenario**
When typing an incomplete tuple type (with one element), such as `void M(int x, (int a) y)`, two identical errors are reported (one from parsing and one from binding).
This fix ensures that parsing always produces at least two tuple elements (missing ones if necessary) and removes the error from binding stage.

**Bugs this fixes:** 
https://github.com/dotnet/roslyn/issues/14091

**Workarounds, if any**
Type a proper tuple type.

**Risk**
**Performance impact**
Low. This is a very localized change to parsing and binding tuple types.

**Is this a regression from a previous update?**
No.

**Root cause analysis:**
The tests that we had for this situation didn't reach binding stage, so we didn't notice the duplication of errors.

**How was the bug found?**
Dogfooding.

### CODE REVIEW

**Comment by jcouv** [comment]
@gafter What do you think of this change? 

Previously, the parser could return a 1-tuple syntax (including a diagnostic), so the binding for tuple types had to handle that case (which would also require a diagnostic to attached to the `ExtendedErrorTypeSymbol`).
With this change, I am adding missing nodes to ensure that the parser always produces at least a 2-tuple.

**Comment by gafter** [comment]
It is possible that this won't parse as a tuple after the PR that I just submitted. If there is no comma, it should just be parsed as a parenthesized expression.

**Comment by gafter** [comment]
Moreover, it should not be considered a declaration expression in the absence of the comma. I suspect the most local fix is to never parse a declaration expression after an open paren unless there is a following comma.

**Comment by gafter** [comment]
Oh, my mistake; this is for parsing of the tuple *type*. In that case I like the fix. LGTM.

**Comment by CyrusNajmabadi** [comment]
> This fix ensures that parsing always produces at least two tuple elements (missing ones if necessary) 

I'm going to push back mildly against this.  The issue here is that missing elements, and parse errors, tend to have larger impacts up the chain on the IDE side.  We're often very conservative when we encounter code like this, and we shut off features in the fear that we'll screw up code that is syntactically incorrect.

So, in general, we far prefer the parser being lenient and parsing things out when it can, and then having the binder come in and state what the problem is.

That means that for list-like things, supporting 0, 1, or many elements is desirable.  That means that places in the grammar where we only accept some sort of subtype (like 'embedded statement') we prefer if we support a much higher type (like 'statement').  etc. etc.

Note, IMO this also helps in other ways:

1. doing less work in the parser means we get a syntax tree faster.  That helps out with a lot of typing scenarios where we have to wait for an up to date tree.
2. doing less work in the parser makes incremental parsing much simpler.  IIRC, we don't reuse nodes with errors in them. So the less nodes with error, the more reuse and the less reparsing we need to do.

Given any grammar-node, my preference would be that the parser fills it out (if possible), reporting only the following types of errors:

1. A required token was not found in the token stream.
2. A token was found that we have no idea what to do with, and we had to skip it.

Pretty much all other errors would be good to move out of hte parser.

**Comment by jcouv** [comment]
The second commit implements Cyrus's recommendation (no parsing error for parsing tuple type with single element, rely on binding to report the error). 
@gafter Let me know if still ok.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. 
We used to report error for tuple type with less than two elements during both parsing and binding. With this PR, it is only reported during binding.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
I am not comfortable with this change. IMHO, syntax errors should be reported in the parser and it should never produce tuples with one element, an appropriate missing nodes should be injected instead.

**Comment by jcouv** [comment]
@CyrusNajmabadi I'll stop by to discuss.
One factor to consider is that the grammar in the spec doesn't allow for parsing 1-tuples, so using a parsing error makes sense (we can discuss if it should be "missing comma" or the special message we previously used): https://github.com/dotnet/roslyn/blob/master/docs/features/tuples.md#language-grammar-changes

**Comment by jcouv** [comment]
I wasn't able to reach Cyrus. I reverted this PR to be the original change (parser parses 0-tuple and 1-tuple types as 2-tuple types with missing syntax nodes).

**Comment by CyrusNajmabadi** [comment]
As mentioned above, I think that's the wrong strategy to take with the parser. The decision around errors in the parser has a large impact in higher levels. We validated in typescript that removing unnecessary restrictions at parse time had quite a positive impact on the rest of the system. 

As to the argument about grammar. The c# parser already deviates greatly from our overly restrictive grammar. We have regularly gone with the approach of pushing the error out until later. As long as the whole compiler correctly errors, then the implementation should be allowed to move things to whatever layer it feels is appropriate to service all the goals of the Roslyn API. 

As mentioned before, IMO, the ideal state for the parser is that it literally only reports errors when there is a token that it needs that it didn't know what to do without, or if it encounters a token it has no idea what to do with. I.e. not having enough, or having too much. 

For any list case, we should allow 0, 1 or many elements if no other grammar productions would match.  We should avoid parse errors when not necessary. Thanks!

**Comment by CyrusNajmabadi** [comment]
Or, in other words: while the grammar is great for expressing many rules ofthe language in a simple, condensed, format; it should not be used to decide if an error is better placed at the syntactic out semantic level. 

Instead, we should use the syntactic *model* as the guide here.  If we can fit the user's code into the syntactic model (without over or under fitting), then we should push the error off until later. Only if we actually just have skipped or missing tokens to deal with the user code should we have errors reported by the parser.

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi There is no such language concept as a tuple with less than two elements. Making this valid and pushing the compiler to deal with it all over adds an unnecessary risk to the product. If you can demonstrate any real issues with the approach, we can revisit, but at this point I don't think parser relaxation is justified.

**Comment by CyrusNajmabadi** [comment]
By definition, there is no language concept for any of the places where our parser is more relaxed than our grammar :-). But it still works out quite well to make the system simple and keep the parser lightweight while avoiding errors when not necessary. 

As for demonstration of real issues, I would say that examples were already listed. For example, because of the parse error things like formatting will be disabled on such a construct. It will also affect incremental parsing. 

Now, these aren't huge issues at all. But it feels to be like it would just be better to avoid all these issues entirely and stick with this style for the parser overall.

**Comment by AlekseyTs** [comment]
> while avoiding errors when not necessary

I am not sure what are you referring to here. I don't remember us ever trying to postpone syntax errors to binding phase just because it can be done.
Also, I believe these errors are necessary, they are syntax errors (because the code doesn't follow the language grammar, there is no ambiguity), and all syntax errors are supposed to be available from the syntax tree, without forcing clients to fish some of them out from binding errors. We are choosing how to recover from an error while parsing, a well-formed tuple node cannot be created, which means there is an error.

**Comment by gafter** [comment]
> I don't remember us ever trying to postpone syntax errors to binding phase just because it can be done.

Agreed; that isn't the reason we do it. We do it when it simplifies the parser's *implemented* grammar in a way that improves the IDE experience. This seems like a good example of that.

**Comment by AlekseyTs** [comment]
> We do it when it simplifies the parser's implemented grammar in a way that improves the IDE experience. 

Please provide existing examples when we don't report syntax error motivated by improvements in IDE experience.

**Comment by gafter** [comment]
> Please provide existing examples when we don't report syntax error motivated by improvements in IDE experience.

```cs
var x = base;
```

```none
Error: Use of keyword 'base' is not valid in this context
```

The language spec's grammar makes this a syntax error, but we permit it in the parser and diagnose it in the binder while binding it to the base type.

We have many examples added in C# 7, such as where we permit `ref` before an expression, before a type, where we permit declaration expressions, etc. Many such syntactic constraints are implemented as binding-time errors, permitting a better IDE experience for the affected code.

We could accomplish the same thing while giving a syntax error in the parser, as long as we don't add a bunch of "missing" nodes to "make it fit". For example, we could accept `(int)` as a tuple type, with an error either in the parser or in the binder, but avoiding adding "missing" syntax.

**Comment by CyrusNajmabadi** [comment]
There are many examples.  One of the largest is that we basically do not even bother with "B.2.10 Interfaces".  If we followed that grammar we would not *syntactically* allow interface methods to have block bodies.  But, in practice, we don't want to generate a poor tree, or produce syntax errors if the user is doing things like changing a class to an interface.  So we allow interface methods to have these bodies and we report the issues later.

Now, i agree that this philosophy has not been adhered to strongly.  But it's really quite nice for us to do this.  And, unless there is a *strong* reason otherwise, i would prefer we err on hte side of only producing syntax errors when we literally cannot fit the user code into our *syntax model*.  

I've talked with Neal at length about this, and i'd like to help out here by moving existing 'unnecessary' syntax checks higher up the stack when i get a chance.  It really does help the IDE out.  We just haven't had time to really address this fully with the existing codebase.  But i would like to avoid making the same mistakes with *new* code that we add.  Hence why i weighed in here :)

**Comment by CyrusNajmabadi** [comment]
Another way to think about this is as such:

We don't want to be in a world where the decisions Mads makes for the health of the language specification to be dictating how our implementation must behave at specific layers.  Mads may or may not choose to represent language rules as grammar rules.  Sometimes he feels like that is appropriate, sometimes he doesn't.  He should be free to make those types of decisions without it forcing a specific implementation on our part.

A good example of this would be 'throw expressions'.  It's possible that the language spec will be explicit at the grammar level where these expressions are allowed.  But that doesn't mean we want actual parse errors to be produced when someone writes reasonable code where the 'throw expression' is in an improper place.  In that regard, we can treat much of the grammar as just shorthand notation for non-syntactic restrictions.  

In other words, we have the choice and flexibility to act as if Mads made a loose grammar and added additional normative rules restricting things after that (even if he actually decides to encode things in the grammar).

**Comment by CyrusNajmabadi** [comment]
I've mentioned 'incremental parsing' a couple of times.  But i wanted to dive into what the actual issue is there.  In order for incremental parsing to work, it is absolutely necessary that the trees produced by the incremetnal parser are identical (in shape and in diagnostics) to those produced by the normal parser.

The more diagnostics we produce at the parser level, the harder this is too achieve.  Specifically, if the parser produces *different* diagnostics for a language construct depending on the context it is in, then we have to be extremely careful about reusing such nodes as the context may not match while incremental parsing.  

Consider the example i gave earlier about MethodDeclarations in Classes vs in Interfaces.  In a system where we parse MethodDeclarations the same no matter where they occur (i.e. we don't listen to the grammar restrictions) then it's trivial to reuse MethodDeclaration nodes if the user changes a TypeDeclaration node from a class to an interface.  However, if there are specific contextual diagnostics we produce, then we have to be conservative and we will not reuse the node.

By pushing unnecessary grammar diagnostics outside of hte parser, we actually vastly simplify the incremental parsing implementation, and we enable *much* more reuse.  The more diagnostics we push in the parser, the less reuse we can get, and the more complex the implementation has to be because us devs have to figure out all the potential context-changes that would restrict reuse.

--

When we did followed the simplified parsing strategy that i've outlined in TypeScript, we ended up with a massively simple IncrementalParsing implementation.  In fact, there are only *3* syntactic cases where we do not reuse a node even though it would be allowed in the new context.  By keeping unnecessary diagnostics out of the typescript parser, we made life *much* easier, and we made things better at the higher levels of the stack.  It has made updating the TS parser super simple without having to be concerned about subtle contextual incremental parsing cases.

**Comment by CyrusNajmabadi** [comment]
As per a request from Julien:

I wanted to clarify why syntactic errors are such an issue for the IDE.  Essentially, we use syntax errors as a proxy for 'this code is broken, and if we start messing around with it, we're likely going to make it worse'.  This proxy heuristic was borne out of the old LS days when we started getting more aggressive with things like formatting/brace-completion, etc.  We quickly discovered that if we remained aggressive in the presence of broken code, we would end up really screwing things up (i.e. inserting/removing whitespace, moving tokens to different lines, and basically aggravating the user enormously).  We realized that a really good way to deal with this would be to just turn off features in the presence of syntax errors.  In other words, we treat syntax errors as a way of saying "we don't know what's going on, and we shouldn't trust the syntax model (i.e. tree) we got back to drive our decisions".

This makes sense when the code is *truly* broken.  i.e. is just out and out missing tokens that we really must have, or has encountered tokens that we have no idea what to do with.  However, it makes less sense when the syntax model is actually totally fine, and it's more that we've just violated some grammar rule.  To use an above example, if the user has a method with a block-body in an interface, that shouldn't have any impact on our ability to format that code.  On the other hand, if they write somethign like "void Foo (!@#!@#)@!#)", then we really don't have a clue what's going on here, and we shouldn't touch it."

We routinely get bugs from people stating that the IDE side isn't working because of cases like this.  I can't find it now, but recently we had one where things were broken when you didn't have an embedded statement inside an 'if statement'.  That's because the parser produced an error here, and we backed off from messing with the code.

Ideally, since the syntax *model* has no difficulty representing this case, we woudl just produce a good tree, and we wouldn't have to disable higher level features.   The error about embedded statements would simply move to some semantic pass and life would be much nicer.

--

So, we'd like the IDE to always behave this way with missing/skipped tokens.  THey're a great proxy for 'something is SUPER wrong and we can't figure out what's going on at all!".  However, in the 1-tuple case, that's not what's going on.  We can see it's a 1-tuple.  We can represent a 1-tuple in the syntax-model.  So we should not inject missing-nodes here.

--

Finally, I feel highly confident with this position given that we followed it very closely in TypeScript.  It kept things super simple (both for parsing and incremental parsing) and produces fantastic trees for the higher levels to use.  It was also an approach that allowed for a systematic methodology to adding new language constructs.  It was almost never the case that a new language construct would ever warrant a new parse error of any sort.

**Comment by AlekseyTs** [comment]
@gafter None of the examples you gave are about not reporting errors about things that are missing, but required by language grammar. None of them are about not adding "missing" nodes in the parser for things that are required. Any examples like that? So that we would compare apples to apples. The examples you gave are about contracts that are legal in one context and are not legal in other contexts. I agree, sometimes, it is much harder to make parser aware of the context. The case in hand, however, is about syntax that is never valid. There is no context when it would be valid.

**Comment by gafter** [comment]
> None of the examples you gave are about not reporting errors about things that are missing, but required by language grammar. None of them are about not adding "missing" nodes in the parser for things that are required. Any examples like that? 

@AlekseyTs `base` as an expression was my first example. It is an example of something missing according to the language grammar. The grammar requires it to be followed by a dot or open square bracket. There is no context in which `base` is permitted by the grammar an expression without one of those two tokens. Yet our parser does not require that; it is enforced in binding.

> I also do not believe that having a code like that is going to be a common scenario.

While typing a tuple type, there is an intermediate point before the comma is typed where this problem occurs. It occurs each and every time a tuple type is being entered through the IDE in the normal way. So I believe it will be a common scenario.

**Comment by AlekseyTs** [comment]
> While typing a tuple type, there is an intermediate point before the comma is typed where this problem occurs. It occurs each and every time a tuple type is being entered through the IDE in the normal way. So I believe it will be a common scenario.

I am fine with whatever IDE experience is going to be for that split second while the comma is typed. It is not going to be common for the situation to last for more than that split of the second.

**Comment by VSadov** [comment]
Typing a tuple type will have to go through error states. For example `(int x, )` is an error since there is a missing element that is not optional, and having missing nodes should normally result in syntax with errors. 
 
The cases that accept incorrect syntax in the parser and then compensate in the binder are typically the cases where validation is hard to do at the parser level. This does not look like one of those cases, so rejecting early seems right. - No need to even try binding this and then solving an issue with having to represent a bad syntax with a bad bound node...

**Comment by jcouv** [comment]
@AlekseyTs @VSadov Thanks for the review. But if that's ok, I will update the PR to add a similar fix for tuple literals (there is indeed a similar situation, as Aleksey hinted). Also, I realized I forgot to check VB (I'll check and add tests if needed).

**Comment by AlekseyTs** [comment]
Sounds good, let's cover all the bases.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler I updated the PR with similar change for tuple expressions and also VB. Thanks

**Comment by jcouv** [comment]
Note: this PR also includes some fixes to test classes so that parser tests would again be able to print out the expected tree.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Since I made further changes (tuple expressions in C# and also corresponding changes in VB), I removed previous approvals. Please review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 1]
```diff
@@ -475,8 +475,7 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
 
             if (typesArray.Length < 2)
             {
-                elementNames?.Free();
-                return new ExtendedErrorTypeSymbol(this.Compilation.Assembly.GlobalNamespace, LookupResultKind.NotCreatable, diagnostics.Add(ErrorCode.ERR_TupleTooFewElements, syntax.Location));
+                throw ExceptionUtilities.Unreachable;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Perhaps `ExceptionUtilities.UnexpectedValue(typesArray.Length)`.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -6693,14 +6693,20 @@ private TupleTypeSyntax ParseTupleType()
                     }
                 }
 
-                var close = this.EatToken(SyntaxKind.CloseParenToken);
-                var result = _syntaxFactory.TupleType(open, list, close);
-
+                if (list.Count < 1)
+                {
+                    list.Add(_syntaxFactory.TupleElement(this.CreateMissingIdentifierName(), identifier: null));
+                }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: `if` can be moved inside `if` below.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 3]
```diff
@@ -3774,12 +3774,12 @@ class C
...
                 Diagnostic(ErrorCode.ERR_InvalidMemberDecl, "=").WithArguments("=").WithLocation(4, 16),
-                // (4,18): error CS8124: Tuple must contain at least two elements.
-                //     var (x, y) = (1, 2);
-                Diagnostic(ErrorCode.ERR_TupleTooFewElements, "(").WithLocation(4, 18),
                 // (4,19): error CS1031: Type expected
                 //     var (x, y) = (1, 2);
                 Diagnostic(ErrorCode.ERR_TypeExpected, "1").WithLocation(4, 19),
+                // (4,19): error CS8124: Tuple must contain at least two elements.
+                //     var (x, y) = (1, 2);
+                Diagnostic(ErrorCode.ERR_TupleTooFewElements, "1").WithLocation(4, 19),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: This error seems confusing since the tuple contains two elements.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Yes. The parser tries to parse `(1, 2)` as a tuple type, but can't, so we get this error.

```C#

        [Fact]
        public void Test()
        {
            var tree = UsingTree(@"
class C
{
    var (x, y) = (1, 2);
}", options: TestOptions.Regular);
            N(SyntaxKind.CompilationUnit);
            {
                N(SyntaxKind.ClassDeclaration);
                {
                    N(SyntaxKind.ClassKeyword);
                    N(SyntaxKind.IdentifierToken, "C");
                    N(SyntaxKind.OpenBraceToken);
                    N(SyntaxKind.MethodDeclaration);
                    {
                        M(SyntaxKind.PredefinedType);
                        {
                            M(SyntaxKind.VoidKeyword);
                        }
                        N(SyntaxKind.IdentifierToken, "var");
                        N(SyntaxKind.ParameterList);
                        {
                            N(SyntaxKind.OpenParenToken);
                            N(SyntaxKind.Parameter);
                            {
                                N(SyntaxKind.IdentifierName);
                                {
                                    N(SyntaxKind.IdentifierToken, "x");
                                }
                                M(SyntaxKind.IdentifierToken);
                            }
                            N(SyntaxKind.CommaToken);
                            N(SyntaxKind.Parameter);
                            {
                                N(SyntaxKind.IdentifierName);
                                {
                                    N(SyntaxKind.IdentifierToken, "y");
                                }
                                M(SyntaxKind.IdentifierToken);
                            }
                            N(SyntaxKind.CloseParenToken);
                        }
                        M(SyntaxKind.SemicolonToken);
                    }
                    N(SyntaxKind.IncompleteMember);
                    {
                        N(SyntaxKind.TupleType);
                        {
                            N(SyntaxKind.OpenParenToken);
                            M(SyntaxKind.TupleElement);
                            {
                                M(SyntaxKind.IdentifierName);
                                {
                                    M(SyntaxKind.IdentifierToken);
                                }
                            }
                            M(SyntaxKind.CommaToken);
                            M(SyntaxKind.TupleElement);
                            {
                                M(SyntaxKind.IdentifierName);
                                {
                                    M(SyntaxKind.IdentifierToken);
                                }
                            }
                            M(SyntaxKind.CloseParenToken);
                        }
                    }
                    N(SyntaxKind.CloseBraceToken);
                }
                N(SyntaxKind.EndOfFileToken);
            }
            EOF();
        }
```

**Comment by cston** [comment]
LGTM

**Comment by jcouv** [comment]
@AlekseyTs @VSadov May I have a second review? Thanks

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -10374,14 +10381,16 @@ private TupleExpressionSyntax ParseTupleExpressionTail(SyntaxToken openParen, Ar
...
 
-                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
-                var result = _syntaxFactory.TupleExpression(openParen, list, closeParen);
-
                 if (list.Count < 2)
                 {
-                    result = this.AddError(result, ErrorCode.ERR_TupleTooFewElements);
+                    list.AddSeparator(SyntaxFactory.MissingToken(SyntaxKind.CommaToken));
+                    var missingExpression = this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_TupleTooFewElements);
+                    list.Add(_syntaxFactory.Argument(nameColon: null, refOrOutKeyword: default(SyntaxToken), expression: missingExpression));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >nameColon: null [](start = 53, length = 15)

Is it valid for the second item to have no name when the first one is named? Let's test scenario when the only element is named to make sure no strange errors reported for the element that we are adding here. Same for tuple type.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I see that we already have tests like this.

---
In reply to: [94264526](https://github.com/dotnet/roslyn/pull/15969#discussion_r94264526) [](ancestors = 94264526)

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Yes, it is valid.
Test `TupleExpressionWithTooFewElements` has tuple expression `(Alice: 1)` and test `TupleTypeWithTooFewElements` has tuple type `(int a)`.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb [hunk 5]
```diff
@@ -757,9 +757,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Dim locationsArray As ImmutableArray(Of Location) = locations.ToImmutableAndFree()
 
                 If typesArray.Length < 2 Then
-                    elementNames?.Free()
-                    diagnostics.Add(ERRID.ERR_TupleTooFewElements, syntax.GetLocation)
-                    Return ErrorTypeSymbol.UnknownResultType
+                    Throw ExceptionUtilities.Unreachable
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb]: >Throw ExceptionUtilities.Unreachable [](start = 20, length = 36)

Consider using ```throw ExceptionUtilities.UnexpectedValue(typesArray.Length)``` for consistency with C#.

File: src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb [hunk 6]
```diff
@@ -1291,6 +1291,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             Dim closeParen As PunctuationSyntax = Nothing
             TryEatNewLineAndGetToken(SyntaxKind.CloseParenToken, closeParen, createIfMissing:=True)
 
+            If argumentBuilder.Count < 2 Then
+                argumentBuilder.AddSeparator(InternalSyntaxFactory.MissingToken(SyntaxKind.CommaToken))
+
+                Dim missingExpression = SyntaxFactory.SimpleArgument(nameColonEquals:=Nothing, expression:=SyntaxFactory.IdentifierName(InternalSyntaxFactory.MissingIdentifier()))
+                missingExpression = ReportSyntaxError(missingExpression, ERRID.ERR_TupleTooFewElements)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Parser/ParseExpression.vb]: >missingExpression = ReportSyntaxError(missingExpression, ERRID.ERR_TupleTooFewElements) [](start = 16, length = 87)

This looks inconsistent with C# implementation, which attaches the error to the MissingIdentifier.

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 7]
```diff
@@ -3019,6 +3019,14 @@ checkNullable:
                 closeParen = closeParen.AddLeadingSyntax(unexpected)
             End If
 
+            If elementBuilder.Count < 2 Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: >If elementBuilder.Count < 2 Then [](start = 12, length = 32)

Consider adding an assert that we never have an empty elementBuilder at this point.

File: src/Compilers/VisualBasic/Portable/Parser/Parser.vb [hunk 8]
```diff
@@ -3019,6 +3019,14 @@ checkNullable:
...
+                elementBuilder.AddSeparator(InternalSyntaxFactory.MissingToken(SyntaxKind.CommaToken))
+
+                Dim missingElement = _syntaxFactory.TypedTupleElement(SyntaxFactory.IdentifierName(InternalSyntaxFactory.MissingIdentifier()))
+                missingElement = ReportSyntaxError(missingElement, ERRID.ERR_TupleTooFewElements)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Parser/Parser.vb]: >missingElement = ReportSyntaxError(missingElement, ERRID.ERR_TupleTooFewElements) [](start = 16, length = 81)

Similar inconsistency as with tuple literal?

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/Parenthesized.vb [hunk 9]
```diff
@@ -138,10 +138,13 @@ BC30451: 'C2' is not declared. It may be inaccessible due to its protection leve
                                  ~~
 BC37259: Tuple must contain at least two elements.
     System.Console.WriteLine(GetType((System.String)))
-                                     ~~~~~~~~~~~~~~~
+                                                   ~
+BC30456: 'C2' is not a member of 'Type'.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/Parenthesized.vb]: This error looks strange. What is the ```'Type'```?

**Comment by AlekseyTs** [comment]
LGTM, modulo the issues 
- https://github.com/dotnet/roslyn/pull/15969#pullrequestreview-14795916
- https://github.com/dotnet/roslyn/pull/15969#pullrequestreview-14795988
- https://github.com/dotnet/roslyn/pull/15969#pullrequestreview-14796089
- https://github.com/dotnet/roslyn/pull/15969#pullrequestreview-14796137
- https://github.com/dotnet/roslyn/pull/15969#pullrequestreview-14796287

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/Parenthesized.vb (hunk 9 above)
Comment (jcouv) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/Parenthesized.vb]: `GetType((C1).C2` is parsed as a member access with the expression being `GetType((C1)` and the member being `C2`. There is a missing token for the close-paren on `GetType` and `(C1)` is a 1-tuple type.

`GetType` returns a `System.Type`. That's the `Type`.

Previously, binding `(C1)` would produce a `ErrorTypeSymbol.UnknownResultType` so I think the binding would not continue. But now, we produce a tuple `(C1, ?)` which allows binding to continue.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by jcouv** [comment]
@srivatsn Please approve for RC.3. This fixes an issue with duplicate errors being reported on malformed tuples. Details in the description.

**Comment by AdamSpeight2008** [comment]
@jcouv 
Could you clarify what happens in the following cases, in the "old" and "new" parser, when they are parse as tuples.
```
() ' 0-Tuple
(0) ' 1-Tuple
(1,2) ' 2-Tuple
```
Would they result in a invalid parse in the "old", and a valid parse in the "new" (but with invalid / unexpected values in their properties. eg the Arguments of said tuple.

**Comment by AdamSpeight2008** [comment]
@jcouv That makes sense, thank you.

**Comment by gafter** [comment]
Actually, I expect and hope that `(0)` as an expression just parses as an ordinary parenthesized expression, and `()` just parses as a parenthsized expression containing a missing identifier.

**Comment by jcouv** [comment]
@gafter Yes, those examples (without tuple names) still get parsed as parenthesized expressions, not tuples.

**Comment by mcintyre321** [comment]
Hi, apologies if this is the wrong place to ask. Is this feature ("Tuple must contain at least two elements") finalised? 

I think it will be a bit of a pain, as removing away an unused  element will require the whole signature caller to change, and calling code to be modified, even if you know you might make it a tuple again later.

Is there a discussion somewhere where the need for this requirement is explained? thanks

**Comment by jcouv** [comment]
@mcintyre321
Maybe I didn't understand your question. The fix above does not change what the C# language allows or doesn't allow. The language spec says that tuples must have at least two elements.
What was changed is the shape of the parse tree that results from parsing bad code? What nodes are produced and where are the errors attached?
From the perspective of a user of a language, this is invisible.
From the perspective of an analyzer write, the parse tree after this fix is not unlike other parse trees for existing bad code.

If this didn't clarify, can you elaborate on your scenario?

**Comment by mcintyre321** [comment]
@jcouv I'm just interested in finding out why the decision was made that single element tuples are disallowed, and thought someone in this thread might be able to point me to the reasoning behind it.

I appreciate they are only tangentially related, but there does exist a `System.Tuple<T>` class, so at some point people thought that was a useful concept, and I myself have used it in the past. I'd like to be able to return single valued tuples, as it will allow me to delete a second return value without having to make lots and lots of changes to any calling code.

**Comment by CyrusNajmabadi** [comment]
Single-valued tuples are only not supported *syntactically*.  You can still write:

```c#
ValueTuple<int> Foo() { return ValueTuple.Create(1); }
```

What we do not support any built-in syntax for this purpose.  

> I'm just interested in finding out why the decision was made

The reason is because we felt it was too problematic to try to deal with the ambiguities present between constructs like single-element-tuples and constructs like casts.  We did not say that we would never support this.  We simply said that we weren't going to invest in *syntax* for single element tuples in C# 7.

**Comment by mcintyre321** [comment]
OK, thanks, that definitely lets me work round having to modify all the callers. I can see how the new syntax might overlap with casting.

**Comment by mcintyre321** [comment]
@CyrusNajmabadi do you think this ("the choice not to allow single valued tuples") will be revisited any time soon? I keep having to use a dummy parameter in certain situations e.g. `class SomeCommand : Command<(string someArgument, object _)> { ... }`

**Comment by CyrusNajmabadi** [comment]
@mcintyre321 https://github.com/dotnet/csharplang/issues/883 appears to be the place where this would be discussed.
